rust/processor/src/worker.rs

```diff
@@ -298,44 +301,42 @@ impl Worker {
         // Create a gap detector task that will panic if there is a gap in the processing
         let (gap_detector_sender, gap_detector_receiver) =
             kanal::bounded_async::<ProcessingResult>(BUFFER_SIZE);
-        let (processor, gap_detection_batch_size, gap_detector_sender) =
-            if self.processor_config.is_parquet_processor() {
-                let processor = build_processor(
-                    &self.processor_config,
-                    self.per_table_chunk_sizes.clone(),
-                    self.deprecated_tables,
-                    self.db_pool.clone(),
-                    Some(gap_detector_sender.clone()),
-                );
-                let gap_detection_batch_size: u64 = self.parquet_gap_detection_batch_size;
+        let is_parquet_processor = self.processor_config.is_parquet_processor();
+        let (processor, gap_detection_batch_size) = if is_parquet_processor {
+            let processor = build_processor(
+                &self.processor_config,
+                self.per_table_chunk_sizes.clone(),
+                self.deprecated_tables,
+                self.db_pool.clone(),
+                Some(gap_detector_sender.clone()),
+            );
+            let gap_detection_batch_size: u64 = self.parquet_gap_detection_batch_size;
+
+            (processor, gap_detection_batch_size)
+        } else {
+            let processor = build_processor(
+                &self.processor_config,
+                self.per_table_chunk_sizes.clone(),
+                self.deprecated_tables,
+                self.db_pool.clone(),
+                None,
+            );
+            let gap_detection_batch_size = self.gap_detection_batch_size;
 
-                (
-                    processor,
-                    gap_detection_batch_size,
-                    Some(gap_detector_sender),
-                )
-            } else {
-                let processor = build_processor(
-                    &self.processor_config,
-                    self.per_table_chunk_sizes.clone(),
-                    self.deprecated_tables,
-                    self.db_pool.clone(),
-                    None,
-                );
-                let gap_detection_batch_size = self.gap_detection_batch_size;
+            (processor, gap_detection_batch_size)
+        };
 
-                (
-                    processor,
-                    gap_detection_batch_size,
-                    Some(gap_detector_sender),
-                )
-            };
+        let gap_detector: Box<dyn GapDetectorTrait + Send> = if is_parquet_processor {
+            Box::new(ParquetFileGapDetector::new(starting_version))
+        } else {
+            Box::new(DefaultGapDetector::new(starting_version))
+        };
```

Comment:
Rather than using a box, make use of enum_dispatch on GapDetector. That way you can just pass around GapDetector instead of a trait object. Nicer types and more efficient.

Reply:
Used GapDetector instead of a trait object